Add WebDAV media backend (#1141)
hedgedoc/hedgedoc: docs/content/media/webdav.md

# WebDAV

You can use any [WebDAV](https://en.wikipedia.org/wiki/WebDAV) server to handle your image uploads in HedgeDoc.

The WebDAV server must host the files in a way that allows HedgeDoc to request and receive them.

You just add the following lines to your configuration:  
<small>(with the appropriate substitution for `<CONNECTION_STRING>`, `<UPLOAD_DIR>`, and `<PUBLIC_URL>` of course)</small>
```
HD_MEDIA_BACKEND="webdav"
HD_MEDIA_BACKEND_WEBDAV_CONNECTION_STRING="<CONNECTION_STRING>"
HD_MEDIA_BACKEND_WEBDAV_UPLOAD_DIR="<UPLOAD_DIR>"
HD_MEDIA_BACKEND_WEBDAV_PUBLIC_URL="<PUBLIC_URL>"
```

The `<CONNECTION_STRING>` should include the username and password (if needed) in the familiar way of `schema://user:password@url`.

With `<UPLOAD_DIR>` you can specify a folder you want to upload to, but you can also omit this, if you prefer to upload directly to the root of the WebDAV server.

Finally, `<PUBLIC_URL>` specifies with which url HedgeDoc can access the upload. For this purpose the filename will be appended to `<PUBLIC_URL>`. So the file `test.png` with `<PUBLIC_URL>` `https://dav.example.com` should be accessible via `https://dav.example.com/test.png`.

## Using NextCloud

> If you want to use the WebDAV server of a NextCloud server, follow the following instructions:

Reviewer: ```suggestion
If you want to use the NextCloud as WebDAV server, follow the following instructions:
```

Author: done

This guide was written using NextCloud 21 in April 2021.

Because the username and password will be included in the config, we suggest using a dedicated NextCloud user for the uploads.  

> In this example the username will be `TestUser` with the password `passw0rd`.

Reviewer 2: Instead of a password you should probably create an app password. You Can find that in your account settings > security.

Author: done

1. In the Files app [create a new folder](https://docs.nextcloud.com/server/latest/user_manual/en/files/access_webgui.html#creating-or-uploading-files-and-directories) that will hold your uploads (e.g `HedgeDoc`).
2. [Share](https://docs.nextcloud.com/server/latest/user_manual/en/files/sharing.html#public-link-shares) the newly created folder. The folder should (per default) be configured with the option `Read Only` (which we will assume in this guide), but `Allow upload and editing` should be fine, too.
3. Get the public link of the share. It should be in your clipboard after creation. If not you can copy it by clicking the clipboard icon at the end of the line of `Share link`. We'll assume it is `https://cloud.example.com/s/some-id` in the following.
4. Append `/download?path=%2F&files=` to this URL. To continue with our example the url should now be `https://cloud.example.com/s/some-id/download?path=%2F&files=`.

> 5. Get the [WebDAV url of you NextCloud server](https://docs.nextcloud.com/server/latest/user_manual/en/files/access_webdav.html). It should be located in the Files app in the bottom left corner under `Settings` > `WebDAV`. Well assume it is `https://cloud.example.com/remote.php/dav/files/TestUser/` in the following.

Reviewer: ```suggestion
5. Get the [WebDAV url of you NextCloud server](https://docs.nextcloud.com/server/latest/user_manual/en/files/access_webdav.html). It should be located in the Files app in the bottom left corner under `Settings` > `WebDAV`. We'll assume it is `https://cloud.example.com/remote.php/dav/files/TestUser/` in the following.
```

Author: done

# WebDAV

You can use any [WebDAV](https://en.wikipedia.org/wiki/WebDAV) server to handle your image uploads in HedgeDoc.

The WebDAV server must host the files in a way that allows HedgeDoc to request and receive them.

You just add the following lines to your configuration:  
<small>(with the appropriate substitution for `<CONNECTION_STRING>`, `<UPLOAD_DIR>`, and `<PUBLIC_URL>` of course)</small>
```
HD_MEDIA_BACKEND="webdav"
HD_MEDIA_BACKEND_WEBDAV_CONNECTION_STRING="<CONNECTION_STRING>"
HD_MEDIA_BACKEND_WEBDAV_UPLOAD_DIR="<UPLOAD_DIR>"
HD_MEDIA_BACKEND_WEBDAV_PUBLIC_URL="<PUBLIC_URL>"
```

The `<CONNECTION_STRING>` should include the username and password (if needed) in the familiar way of `schema://user:password@url`.

> With `<UPLOAD_DIR>` you can specify a folder you want to upload to, but you can also omit this, if you prefer to upload directly to the root of the WebDAV server.

Reviewer 3: Omitting means setting `HD_MEDIA_BACKEND_WEBDAV_UPLOAD_DIR` so an empty string of not setting it?

Author: clarified

Finally, `<PUBLIC_URL>` specifies with which url HedgeDoc can access the upload. For this purpose the filename will be appended to `<PUBLIC_URL>`. So the file `test.png` with `<PUBLIC_URL>` `https://dav.example.com` should be accessible via `https://dav.example.com/test.png`.

> ## Using NextCloud

Reviewer 3: The 'c' in 'Nextcloud' is *not* capitalized.

Author: fixed
